### PR TITLE
add edition aliases for minimum toolchain

### PR DESCRIPTION
**Pull request by danieleades**

### CODE REVIEW

**Comment by foresterre** [comment]
> No worries, it probably will practically have the same effect.
> 
> An advantage of #104 will be that the there can be no semantically invalid search space since the edition is just an alias for a given version. If however given `--edition 2018 --minimum 1.10.0`, we would need to ensure the smallest possible search space would be used, i.e. in case of the example, the specified minimum is smaller than the edition, so we'll need to override the minimum to be the minimum version supported by this edition, i.e. `1.31.0`, even though the argument clearly shows 'minumum=1.10.0`, potentially leading to confusion.
> 
> A question I ask myself though is whether it is the same thing to say specifying an edition as simply an alias for specifying the version in which the edition was introduced; i.e. whether an edition option (for cargo-msrv) might introduce additional (side) effects aside from limiting the version search space. I'm not yet sure whether it would.
> 
> A disadvantage of #104 could be that the option would accept too many different types of values, making the interface less understandable. As long as the edition is simply an alias for the first Rust version which adds support for the edition, I don't think it would be a problem though.

_Originally posted by @foresterre in https://github.com/foresterre/cargo-msrv/issues/96#issuecomment-927167467_

File: src/config.rs [hunk 1]
```diff
@@ -257,3 +257,34 @@ impl<'config> TryFrom<&'config ArgMatches<'config>> for Config<'config> {
         Ok(builder.build())
     }
 }
+
+fn parse_version(input: &str) -> Result<semver::Version, semver::Error> {
+    match input {
+        "2018" => Ok(semver::Version::new(1, 31, 0)),
```
Comment (foresterre) [inline src/config.rs:263]: For completeness sake, we should also add the first edition, i.e. "2015" (version 1.0.0). 

https://doc.rust-lang.org/edition-guide/rust-2015/index.html

**Review by foresterre — CHANGES_REQUESTED** [review]

**Comment by foresterre** [comment]
Thanks for your contributions 😄.

I've left two minor requests for changes, and a more discussion style comment on how to design the interface around editions. 

I think for the purpose of cargo-msrv's `--minimum` option we can have the edition be an alias for the first version which it supports. We should however remain aware, that regardless, even if such a version was the first version to support an edition, it does not mean that specifying an edition will mean we test only for support of the given edition; i.e. the context of the edition alias here is only that it represents the earliest version which supported said edition.

We should be cautious not to introduce confusion because of terminology which could imply more than intended.

Regardless, I think in this case, supporting an alias for the first version of a toolchain which supports the given edition is indeed an acceptable solution for the same problem which #104 aimed to solve. After the two minor issues have been addressed I will be happy to merge this PR.

File: src/config.rs (hunk 1 above)
Comment (danieleades) [reply src/config.rs:264]: done!

File: src/cli.rs [hunk 2]
```diff
@@ -79,7 +79,7 @@ so: `rustup run <toolchain> <COMMAND...>`. You'll only need to provide the <COMM
                     .visible_alias("minimum")
                     .help("Earliest version to take into account")
                     .long_help("Earliest (least recent) version to take into account. \
-                     Version must match a valid Rust toolchain, and be semver compatible.")
+                     Version must match a valid Rust toolchain, and be semver compatible. Edition aliases may also be used")
```
[the remark this answers is not in the document]
Comment (danieleades) [reply src/cli.rs]: done!

**Comment by danieleades** [comment]
i created this pull request before i saw you had #96, and as per your comment in that PR, they're not trivially compatible flags. However i think being able to set a minimum has greater utility than setting a range, since a later compiler can still compile an earlier edition.

for example a compiler version that came out after 2018 edition stabilised can still compile a 2015 edition crate. it's the lower bound that we're interested in, and the upper bound doesn't matter so much (for the use cases i'm thinking of, at least)

**Review by foresterre — APPROVED** [review]

**Pull request merged**